src/backend/distributed/connection/connection_management.c

```diff
@@ -310,25 +311,86 @@ StartNodeUserDatabaseConnection(uint32 flags, const char *hostname, int32 port,
 	if (!(flags & FORCE_NEW_CONNECTION))
 	{
 		/* check connection cache for a connection that's not already in use */
-		connection = FindAvailableConnection(entry->connections, flags);
+		MultiConnection *connection = FindAvailableConnection(entry->connections, flags);
 		if (connection)
 		{
 			return connection;
 		}
 	}
 
+
+	if (flags & NEVER_WAIT_FOR_CONNECTION)
+	{
+		/*
+		 * The caller doesn't want the connection manager to wait
+		 * until a connection slot is available on the remote node.
+		 * In the end, we might fail to establish connection to the
+		 * remote node as it might not have any space in
+		 * max_connections for this connection establishment.
+		 *
+		 * Still, we keep track of the connnection counter.
+		 */
+		IncrementSharedConnectionCounter(hostname, port);
+	}
+	else if (flags & OPTIONAL_CONNECTION)
```

Comment:
do we need to do some sanity checks on setting multiple conflicting flags?

Reply:
Cant we programmatically prevent setting two flags at the same time ? If we used a method to set the flags, we could check it inside that method easily.

Reply:
Actually as a general suggest I think we should always use methods for these things which have checks to prevent incorrect state. Otherwise in the future someone else or even the author can set both of the conflicting flags by a mistake. Asserts are sanity checks but programmatically adding these checks will help us avoid these problems.

Reply:
I already added asserts, in my point of view other ways of doing adds some level complexity and extra code, which is also something to be careful about. As this is a very simple check/situation, I doubt it'd help to introduce new function(s). For more complicated things,  like you've done on TaskQuery, that sounds viable/desirable.

Reply:
If someone later sets conflicting flags in some other part but they don't have assertions enabled. Isn't there a change that we will merge that change?

Reply:
First, I always develop with asserts enabled, before sometimes also run tests without asserts enabled. And, I encourage everyone to do that as well :) Second, we have an item in the release tests that explicitly asks to run the regression tests w/wout asserts enabled. So, even  if we merge that  change, there is a very little chance that it'll be on the release. If we release without running the tests asserts enabled, then it's a really problem I think.

We should re-consider running CI with asserts enabled as well, at least some schedules.

Anyway, if you have strong opinions on this, feel free to open a PR on top of the PR, and we can discuss the trade-offs on the code.

Reply:
We can have nightly job where we run the tests with asserts. However we should first fix the flaky tests I think..

Also my suggestion was this: https://github.com/citusdata/citus/commit/ea09b3309b2e219ba7bc84a38dffe13ed419197c